classes/Link.php

```diff
@@ -772,7 +773,86 @@ public function getAdminLink($controller, $withToken = true, $sfRouteParams = ar
 
         $idLang = Context::getContext()->language->id;
 
-        return $this->getBaseLink().basename(_PS_ADMIN_DIR_).'/'.Dispatcher::getInstance()->createUrl($controller, $idLang, $params);
+        return $this->getAdminBaseLink().basename(_PS_ADMIN_DIR_).'/'.Dispatcher::getInstance()->createUrl($controller, $idLang, $params);
+    }
+
+    /**
+     * @param int|null  $idShop
+     * @param bool|null $ssl
+     * @param bool      $relativeProtocol
+     * @return string
+     * @throws PrestaShopDatabaseException
+     */
+    public function getAdminBaseLink($idShop = null, $ssl = null, $relativeProtocol = false)
+    {
+        static $force_ssl = null;
+
+        if ($ssl === null) {
+            if ($force_ssl === null) {
+                $force_ssl = (Configuration::get('PS_SSL_ENABLED') && Configuration::get('PS_SSL_ENABLED_EVERYWHERE'));
+            }
+            $ssl = $force_ssl;
+        }
+
+        if (Configuration::get('PS_MULTISHOP_FEATURE_ACTIVE')) {
+            if (null === $idShop) {
+                $idShop = $this->getMatchingUrlShopId();
+            }
+
+            //Use the matching shop if present, or fallback on the default one
+            if (null !== $idShop) {
+                $shop = new Shop($idShop);
+            } else {
+                $shop = new Shop(Configuration::get('PS_SHOP_DEFAULT'));
+            }
+        } else {
+            $shop = Context::getContext()->shop;
+        }
+
+        if ($relativeProtocol) {
+            $base = '//'.($ssl && $this->ssl_enable ? $shop->domain_ssl : $shop->domain);
+        } else {
+            $base = (($ssl && $this->ssl_enable) ? 'https://'.$shop->domain_ssl : 'http://'.$shop->domain);
+        }
+
+        return $base.$shop->getBaseURI();
+    }
+
+    /**
+     * Search for a shop whose domain matches the current url
+     *
+     * @return int|null
+     */
+    public function getMatchingUrlShopId()
+    {
+        if (null === $this->urlShopId) {
+            $host = Tools::getHttpHost();
+            $request_uri = rawurldecode($_SERVER['REQUEST_URI']);
+
+            $sql = 'SELECT s.id_shop, CONCAT(su.physical_uri, su.virtual_uri) AS uri, su.domain, su.main
+                    FROM '._DB_PREFIX_.'shop_url su
+                    LEFT JOIN '._DB_PREFIX_.'shop s ON (s.id_shop = su.id_shop)
+                    WHERE (su.domain = \''.pSQL($host).'\' OR su.domain_ssl = \''.pSQL($host).'\')
+                        AND s.active = 1
+                        AND s.deleted = 0
+                    ORDER BY LENGTH(CONCAT(su.physical_uri, su.virtual_uri)) DESC';
+
+            try {
+                $result = Db::getInstance()->executeS($sql);
+            } catch (PrestaShopDatabaseException $e) {
+                $result = array();
+            }
+
+            foreach ($result as $row) {
+                // A shop matching current URL was found
+                if (preg_match('#^'.preg_quote($row['uri'], '#').'#i', $request_uri)) {
```

Comment:
What kind of uri do you get in `$row`? 
You only check if the request uri starts with `$row['uri']`, can be false if you have a request uri: `test.com/product/beer/awesome` and `$row['uri']`: test.com/project/beer

Reply:
$row['uri'] contains the shop domain concatenated to the virtual part, so it will contain test.com/my_virtual_folder wich will be the base of all the urls of this shop context, thus you will have urls like test.com/my_virtual_folder/product/beer/awesome and test.com/my_virtual_folder/project/beer which will all match this uri
I the url does not match it means you are in another shop context